codejail/jail_code.py

```diff
@@ -240,9 +270,12 @@ def jail_code(command, code=None, files=None, extra_files=None, argv=None,
         # Add the code-specific command line pieces.
         cmd.extend(argv)
 
+        # Determine effective resource limits.
+        effective_limits = get_effective_limits(limit_overrides_context)
+
         # Use the configuration and maybe an environment variable to determine
         # whether to use a proxy process.
-        use_proxy = LIMITS["PROXY"]
+        use_proxy = effective_limits["PROXY"]
```

Comment:
I understand the symmetry of using effective_limits for PROXY, but I'm not sure it makes sense to switch this on and off per-course.  Of course, if we never override PROXY, then it doesn't make a difference.

Reply:
@nedbat I'd like to keep the symmetry of using effective_limits everywhere, but if you think it's worth guarding against, I could add a check to override_settings that raises an exception if one tries to override the value of PROXY.

Reply:
My first question was whether it would work to let PROXY differ between invocations, and I think the answer is yes.  

The original reason for the proxy setting was to deal with memory issues: when a subprocess is spawned, it is a copy of the parent process, so if the parent is using more than 50% of memory, the subprocess could fail because it can't get enough resources to duplicate the parent.

The proxy process was added so that it could be small, and then it could create subprocesses without fear of failing.

We run with PROXY=1 in production. I can't imagine a reason for overriding it, and if we did, it would likely only cause problems.

But, symmetry is good.

BTW: I'm disappointed to see that I did not write an explanation of PROXY like this in the codejail repo.... :(

Reply:
@nedbat Makes sense. What I'm proposing is adding a guard like:
```
def override_limit(limit_name, value, limit_overrides_context):
    if limit_name == 'PROXY':
        log.error('PROXY toggle may not be overriden on a per-context basis; it must be set globally.')
        return
    if limit_overrides_context not in LIMIT_OVERRIDES:
        LIMIT_OVERRIDES[limit_overrides_context] = {}
    LIMIT_OVERRIDES[limit_overrides_context][limit_name] = value
```

allowing us to use `effective_limits` everywhere, but guarding against inconsistent proxy settings. What do you think?

Reply:
^ I pushed this change.